File: cpp/56.merge-intervals.cpp
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> merge(vector<vector<int>> &intervals) {
    sort(intervals.begin(), intervals.end(),
         [](const vector<int> &a, const vector<int> &b) -> bool {
           return a[0] < b[0];
         });
    vector<vector<int>> ans;
    int cur_begin = intervals[0][0];
    int cur_end = intervals[0][1];
    for (auto interval : intervals) {
      if (interval[0] <= cur_end) {
        cur_end = max(cur_end, interval[1]);
      } else {
        ans.push_back({cur_begin, cur_end});
        cur_begin = interval[0];
        cur_end = interval[1];
      }
    }
    ans.push_back({cur_begin, cur_end});
    return ans;
  }
};

File: py/56.merge-intervals.py
from typing import List


class Solution:
    def merge_1(self, intervals: List[List[int]]) -> List[List[int]]:
        # O(nlogn)
        intervals.sort(key=lambda e: e[0])
        print(intervals)
        ans = []
        cur_start = intervals[0][0]
        cur_end = intervals[0][1]
        for interval in intervals[1:]:
            if interval[0] <= cur_end:
                cur_end = max(cur_end, interval[1])
            else:
                ans.append([cur_start, cur_end])
                cur_start = interval[0]
                cur_end = interval[1]
        ans.append([cur_start, cur_end])
        return ans

        # Follow-up: no sorting
